src/music/midifile.cpp

```diff
@@ -450,8 +794,304 @@ void MidiFile::MoveFrom(MidiFile &other)
+			case MIDIST_PROGCHG:
+			case MIDIST_CHANPRESS:
+				fwrite(dp, 1, 2, f);
+				dp += 2;
+				continue;
+			}
+
+			/* Sysex needs to measure length and write that as well */
+			if (*dp == MIDIST_SYSEX) {
+				fwrite(dp, 1, 1, f);
+				dp++;
+				byte *sysexend = dp;
+				while (*sysexend++ != MIDIST_ENDSYSEX);
+				ptrdiff_t sysexlen = sysexend - dp;
+				WriteVariableLen(f, sysexlen);
+				fwrite(dp, 1, sysexend - dp, f);
+				dp = sysexend;
+				continue;
+			}
+
+			/* Fail for any other commands */
+			fclose(f);
+			return false;
+		}
+	}
+
+	/* End of track marker */
+	fwrite(&(bb = 0x00), 1, 1, f);
+	fwrite(&(bb = MIDIST_SMF_META), 1, 1, f);
+	fwrite(&(bb = 0x2F), 1, 1, f);
+	fwrite(&(bb = 0x00), 1, 1, f);
+
+	/* Fill out the RIFF block length */
+	size_t trackendpos = ftell(f);
+	fseek(f, tracksizepos, SEEK_SET);
+	uint32 tracksize = trackendpos - tracksizepos - 4;
+	tracksize = TO_BE32(tracksize);
+	fwrite(&tracksize, 4, 1, f);
+
+	fclose(f);
+	return true;
+}
+
+/**
+ * Get the name of a Standard MIDI File for a given song.
+ * For songs already in SMF format, just returns the original.
+ * Otherwise the song is converted and a temporary file written,
+ * and the temporary file's name returned.
+ * @note Caller is responsible for freeing the returned string.
+ * @param song Song definition to query
+ * @return Full filename string, NULL if failed
+ */
+char *MidiFile::GetSMFFile(const MusicSongInfo &song)
+{
+	if (song.filetype == MTT_STANDARDMIDI) {
+		return stredup(song.filename);
+	}
+
+	if (song.filetype != MTT_MPSMIDI) return NULL;
+
+	const char *lastpathsep = strrchr(song.filename, PATHSEPCHAR);
+	if (lastpathsep == NULL) {
+		lastpathsep = song.filename;
+	}
+
+	char basename[MAX_PATH];
+	{
+		/* Remove all '.' characters from filename */
+		char *wp = basename;
+		for (const char *rp = lastpathsep + 1; *rp != '\0'; rp++) {
+			if (*rp != '.') *wp++ = *rp;
+		}
+		*wp++ = '\0';
+	}
+
+	char tempdirname[MAX_PATH];
+	FioGetFullPath(tempdirname, lastof(tempdirname), Searchpath::SP_AUTODOWNLOAD_DIR, Subdirectory::BASESET_DIR, basename);
+	if (AppendPathSeparator(tempdirname, lastof(tempdirname)) == false) return NULL;
+	FioCreateDirectory(tempdirname);
+
+	char output_filename[MAX_PATH];
+	seprintf(output_filename, lastof(output_filename), "%s%d.mid", tempdirname, (MpsMachine::glitch[0] == 0) ? song.cat_index : -1);
+
+	if (FileExists(output_filename) && MpsMachine::glitch[0] == 0) {
+		/* If the file already exists, assume it's the correct decoded data */
+		return stredup(output_filename);
+	}
+
+	byte *data;
+	size_t datalen;
+	data = GetMusicCatEntryData(song.filename, song.cat_index, datalen);
+	if (data == NULL) return NULL;
+
+	MidiFile midifile;
+	if (!midifile.LoadMpsData(data, datalen)) {
+		free(data);
+		return NULL;
+	}
+	free(data);
+
+	if (midifile.WriteSMF(output_filename)) {
+		return stredup(output_filename);
+	} else {
+		return NULL;
+	}
+}
+
+
+static bool CmdGlitchMusic(byte argc, char *argv[])
```

Comment:
...Why? Why would we want to toggle the "glitchiness", whatever that is (I didn't see it explained anywhere)